### PR TITLE
Adding Consideration Sets Model to Mixed Logit

### PR DESCRIPTION
**Pull request by NathanielF**
## Description
A special case of discrete choice models implement a kind of two-stage model where first choice of goods is "screened" off as worthy of attention/consideration and only then are the utility of each good modelled and used to predict market share. I'm proposing to offer this type of modelling as an extension of the mixed logit class. 

## Related Issue

- [ ] Closes #
- [X ] Related to #
- https://github.com/pymc-labs/pymc-marketing/issues/2443

## Summary

Adds `ConsiderationSetMixedLogit`, a two-stage discrete choice model that separates **consideration** (who gets evaluated) from **preference** (who is chosen conditional on consideration). Extends the existing `MixedLogit` class.

- **New class** (`consideration_set_logit.py`, 542 lines): Implements the bridge formula `P(j|n) = softmax(log(π_nj) + V_nj)` where consideration probabilities π are driven by exclusion-restricted instruments Z_tilde, separate from utility covariates X. Supports single or multi-instrument Z_tilde `(N, J)` or `(N, J, K_z)`, optional alternative-specific consideration intercepts, and optional random consideration intercepts per individual.
- **Test suite** (`test_consideration_set_logit.py`, 587 lines, 22 tests): Covers model construction, config, prior predictive, posterior predictive with counterfactual data, `apply_intervention` for both refit and no-refit branches, multi-instrument variants, numerical stability under extreme Z_tilde, and idata attribute serialisation.
- **Documentation notebook** (`consideration_set_logit.ipynb`): Parameter recovery on simulated hiring data, then empirical application to Swiss Metro travel mode choice with 4 consideration instruments. Includes counterfactual analysis via `apply_intervention`, LOO-CV model comparison, and discussion of identification caveats.
- **Gallery entry** added to `gallery.md`.
- **Public export** added to `customer_choice/__init__.py`.

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://www.pymc-marketing.io/en/latest/contributing/index.html). Feel free to comment [`pre-commit.ci autofix` to auto-fix](https://pre-commit.ci/#configuration-autofix_prs).
- [ X] Included tests that prove the fix is effective or that the new feature works
- [ X] Added necessary documentation (docstrings and/or example notebooks) using [numpydoc format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

----
📚 Documentation preview 📚: https://pymc-marketing--2442.org.readthedocs.build/en/2442/

### CODE REVIEW

**Comment by NathanielF** [comment]
Giving this one another nudge @timbo112711 , @juanitorduz 

If you get a chance this week can you let me know what you think?

**Comment by juanitorduz** [comment]
@NathanielF , here is an initial review from CodeX:

Overview: "Review PR #2442 (excluding notebook) and provide a prioritized implementation plan to address quality and maintainability risks in the new Consideration Set Mixed Logit model."
todos:
  - id: guard-z-dim-switch
    content: Add runtime guards (or safe rebuild path) for incompatible Z_tilde dimensionality changes after model build
    status: pending
  - id: validate-consideration-inputs
    content: Implement strict validation for Z_tilde shape/rank and z_instrument_names length
    status: pending
  - id: add-negative-tests
    content: Add tests for invalid consideration instrument shapes, names, and mismatch scenarios
    status: pending
  - id: tighten-type-hints
    content: Replace generic dict annotations with typed structures for consideration of instrument interfaces
    status: pending
  - id: rng-test-hygiene
    content: Use local RNG objects in tests instead of global np.random.seed mutation
    status: pending
isProject: false
---

# PR 2442 Improvement Plan

## Scope Reviewed
- Code: [pymc_marketing/customer_choice/consideration_set_logit.py](pymc_marketing/customer_choice/consideration_set_logit.py)
- Tests: [tests/customer_choice/test_consideration_set_logit.py](tests/customer_choice/test_consideration_set_logit.py)
- API/export/config/docs touchpoints: [pymc_marketing/customer_choice/__init__.py](pymc_marketing/customer_choice/__init__.py), [pyproject.toml](pyproject.toml), [.pre-commit-config.yaml](.pre-commit-config.yaml), [docs/source/gallery/gallery.md](docs/source/gallery/gallery.md)
- Explicitly excluded: notebook changes.

## Critical
- Enforce consistency between model graph dimensionality and updated `consideration_instruments` during posterior predictive/intervention calls.
  - Add validation in `sample_posterior_predictive` and `apply_intervention` to prevent switching between 2D and 3D `Z_tilde` on an already-built model (or rebuild model safely when dimensionality changes).
  - Fail fast with a clear `ValueError` before `pm.set_data` when shape/dims are incompatible.

## Major
- Add strict input validation for `consideration_instruments`:
  - Validate `Z_tilde` dtype and rank (`ndim in {2, 3}`), expected `(N, J[, K_z])` shape, and alignment with `choice_df` row count and number of alternatives.
  - Validate `z_instrument_names` length equals `K_z` when provided.
- Strengthen tests around failure and edge paths:
  - Add tests that assert informative errors for shape mismatch, missing/incorrect `z_instrument_names`, and row-count mismatch between `choice_df` and `Z_tilde`.
  - Add a test that attempts 2D→3D and 3D→2D updates after fit and asserts intended behavior (reject or rebuild).

## Minor
- Improve API typing and docs consistency:
  - Replace bare `dict` annotations with a typed mapping/`TypedDict` for `consideration_instruments` and related method parameters.
  - Expand docstrings for public methods to match existing project style (include explicit constraints and failure modes).
- Test hygiene:
  - Replace `np.random.seed(42)` global mutation in test data generator with local RNG instance to avoid cross-test coupling.

## Optional
- Modularity refactor:
  - Extract shared prediction-data update logic into a private helper to reduce duplication between `sample_posterior_predictive` and `apply_intervention`.
- Docs polish:
  - Add a short API-level reference note in class/method docstrings for consideration-set identification assumptions.

**Comment by NathanielF** [comment]
Thanks @juanitorduz this is probably enough to get some refinements in. I think they should be straightforward.

**Comment by juanitorduz** [comment]
The notebook looks great!

**Comment by NathanielF** [comment]
I think this should be ok now @juanitorduz , @timbo112711

**Comment by juanitorduz** [comment]
Thanks @NathanielF 

This looks better :) 

I will approve, but before we merge, I suggest (1) adding a `hide-input` tag to the plotting cells in the notebook to hide the code (it is a bit overwhelming and is not necessarily necessary. And (2) act on some suggestions from Opus 4.7 :D (see below)

---

# Review of PR #2442 — Consideration Set Mixed Logit

Measured against the previous [CodeX feedback](https://github.com/pymc-labs/pymc-marketing/pull/2442#issuecomment-4234795860), reviewing the `consideration_sets` branch at `c3b70701`.

## Summary

Most of the critical and major items are addressed: the `TypedDict` refactor, the dimensionality guard, and the new negative tests are clean. A few gaps remain (row/alt shape validation, the optional refactor, a lingering hardcoded seed).

## Item-by-item assessment

### Critical — Z_tilde dimensionality switch guard: ADDRESSED

`_update_consideration_instruments` in `pymc_marketing/customer_choice/consideration_set_logit.py` raises a clear `ValueError` *before* any `pm.set_data` is reached, and both `sample_posterior_predictive` and `apply_intervention` route through it:

```python
if hasattr(self, "model") and self._multi_instrument != new_multi:
    old_dim = "3-D" if self._multi_instrument else "2-D"
    new_dim = "3-D" if new_multi else "2-D"
    raise ValueError(
        f"Cannot switch Z_tilde dimensionality after model has been built "
        f"(built with {old_dim}, new array is {new_dim}). "
        "Rebuild the model with the new dimensionality."
    )
```

Covered by `test_dim_switch_2d_to_3d_after_build_raises` and `test_dim_switch_3d_to_2d_after_build_raises`.

### Major — Strict input validation: PARTIALLY ADDRESSED

Done:
- Rank check (`ndim in {2, 3}`).
- `z_instrument_names` length equals `K_z`.
- Missing `Z_tilde` key.

Gaps (CodeX explicitly asked for these):
- **No alignment check against `choice_df`**: `Z_tilde.shape[0] != len(choice_df)` is never validated. A mismatched row count only surfaces later as a cryptic PyMC shape error.
- **No alignment check against number of alternatives** (`Z_tilde.shape[1] != n_alts`).
- **No dtype check** on `Z_tilde`. Low priority but asked.

Suggested addition inside `_update_consideration_instruments` (after the rank check):

```python
n_rows = len(self.choice_df) if hasattr(self, "choice_df") else None
if n_rows is not None and Z.shape[0] != n_rows:
    raise ValueError(
        f"Z_tilde has {Z.shape[0]} rows but choice_df has {n_rows}."
    )
if not np.issubdtype(Z.dtype, np.number):
    raise ValueError(f"Z_tilde must be numeric, got dtype {Z.dtype}.")
```

Note: at construction time `self.choice_df` isn't set yet (the call happens *before* `super().__init__`), so either (a) move row-count validation to after `super().__init__(...)` in `__init__`, or (b) pass `choice_df` into `_update_consideration_instruments`.

### Major — Negative tests: MOSTLY ADDRESSED

Six new validation tests were added in `TestConsiderationSetMixedLogitValidation`: rank 1D/4D, name length mismatch, missing key, and both 2D→3D and 3D→2D post-build rejections.

Missing (paired with the validation gaps above):
- Row-count mismatch: `Z_tilde` with `shape[0] != len(choice_df)`.
- Alternatives mismatch: `Z_tilde.shape[1] != n_alts` (e.g., 3 alternatives but `Z_tilde` has `J=2`).

### Minor — Type hints: ADDRESSED

A proper `ConsiderationInstruments(TypedDict)` with a required `_ConsiderationInstrumentsRequired` base nicely models the required/optional keys:

```python
class _ConsiderationInstrumentsRequired(TypedDict):
    Z_tilde: np.ndarray


class ConsiderationInstruments(_ConsiderationInstrumentsRequired, total=False):
    """Typed dict for consideration instrument inputs.
    ...
    """

    z_instrument_names: list[str]
```

Propagated through `__init__`, `sample_posterior_predictive`, `apply_intervention`, and `_update_consideration_instruments`.

### Minor — Docstrings: ADDRESSED
The class docstring now has an explicit **Notes** section covering the exclusion restriction, the alias between `gamma_0j` and `alpha_j`, and the eta-mean-zero identification caveat.
### Minor — RNG test hygiene: ADDRESSED (with a small nit)
No more `np.random.seed(42)` global mutation. `_generate_hiring_data` now uses a *local* `local_rng`:
```python
def _generate_hiring_data(n=50, random_consideration=False, seed=42):
    """Generate synthetic hiring data with consideration instruments."""
    local_rng = np.random.default_rng(seed)
```
Nit: the module-level `rng = np.random.default_rng(seed)` is still shared state across the multi-instrument tests. In practice this is fine as long as discovery order doesn't change, but a fully hygienic version would create the rng inside each test body.
### Optional — Modularity refactor: NOT ADDRESSED
The duplication CodeX called out still exists. `sample_posterior_predictive` and the `apply_intervention` no-refit branch both assemble the same `{"X", "y", "W", "Z"}` dict and call `pm.set_data`. A small private helper would reduce the risk of them drifting:
```python
def _build_prediction_data_dict(
    self,
    new_X: np.ndarray | None,
    new_F: np.ndarray | None,
    new_y: np.ndarray | None,
) -> dict[str, np.ndarray]:
    data_dict: dict[str, np.ndarray] = {}
    if new_X is not None and new_y is not None:
        data_dict["X"] = new_X
        data_dict["y"] = new_y
        if new_F is not None and len(new_F) > 0:
            data_dict["W"] = new_F
    data_dict["Z"] = self.consideration_instruments["Z_tilde"]
    return data_dict
```
Low priority, but worth doing while the file is fresh.
## Additional issues found
1. **Hardcoded `random_seed=100` with a TODO** in `apply_intervention`:
    ```python
    idata_new_policy = pm.sample_posterior_predictive(
        self.idata,
        var_names=["p", "likelihood"],
        return_inferencedata=True,
        extend_inferencedata=False,
        random_seed=100,  # TODO: inherit from sampler_config or accept as param
    )
    ```
    TODOs on a random seed shouldn't ship. Either plumb it through `fit_kwargs` / `sampler_config["random_seed"]`, or expose a `random_seed` kwarg on `apply_intervention` (which is what the no-refit branch in `sample_posterior_predictive` does via `**kwargs`).
2. **Possibly unreachable branch in `apply_intervention`**:
    ```python
    if not hasattr(self, "model"):
        self.sample()
    ```
    `self.model` is set by the parent's `build_model()` (called from `fit` / `sample`). Once a user has called `fit`/`sample` once, this is always `True`; if they haven't, `self.idata` is `None` and the subsequent `pm.sample_posterior_predictive(self.idata, ...)` would fail anyway. Either remove the branch or make it robust (call `self.fit(**fit_kwargs)` explicitly with the provided kwargs, and document the side effect).
3. **`test_apply_intervention_with_new_equations` doesn't meaningfully exercise the refit branch.** The "different" utility spec is just the same covariates in swapped additive order (`exp_A + age_A` vs `age_A + exp_A`) — patsy produces an identical design matrix. The test still runs the refit code path, so it's not broken, but it doesn't assert the new utility structure actually changed anything. Consider dropping a covariate, or adding an interaction, to make the refit test semantically meaningful.
4. **Bridge formula numerical stability** (just an observation): using the `log(sigmoid(x)) = x - softplus(x)` identity and then subtracting the row-max before softmax is nicely done:
    ```python
    pi = pm.Deterministic("pi", pm.math.sigmoid(log_odds), dims=("obs", "alts"))
    # Use log-sigmoid identity: log(sigmoid(x)) = x - softplus(x)
    # This is numerically superior to log(sigmoid(x) + eps)
    log_pi = log_odds - pt.softplus(log_odds)
    ```
    The `test_extreme_z_tilde_no_nan` test backs this up at `z = -10`. Consider extending it to `z = -50` to future-proof against changes that reintroduce the `log(pi + eps)` formulation.
5. **`random_consideration=True` without a `group_id`**: `eta = eta_individual` is used directly when `grp_idx is None`, assuming the `"individuals"` coord has length `n_obs`. Worth double-checking the parent `MixedLogit` guarantees that invariant; if yes, a one-line `assert` + comment would help readers who want to audit the identification argument.
## Priority for merge
Must-fix before approval:
- Remove the `random_seed=100` TODO (`apply_intervention`).
- Add row-count / alternative-count validation in `_update_consideration_instruments` and two paired negative tests. This closes the last piece of the CodeX "Major" bucket.
Should-fix (same PR, small):
- Revisit the `if not hasattr(self, "model"): self.sample()` branch in `apply_intervention`.
- Make `test_apply_intervention_with_new_equations` actually change the design matrix.
Nice-to-have (can be follow-up):
- Extract the `_build_prediction_data_dict` helper.
- Move the module-level `rng` in tests into each test body.
- Extend extreme-`Z_tilde` test to `z = -50`.
Overall: the PR is in good shape and the author has clearly worked through CodeX's critical/major list. Once the shape-alignment validation and the hardcoded seed are cleaned up, I'd be comfortable approving.

**Review by juanitorduz — APPROVED** [review]

**Comment by NathanielF** [comment]
Thanks @juanitorduz !
